diff --git a/src/core/mac/sub_mac.hpp b/src/core/mac/sub_mac.hpp
index e334091da..53d72f59f 100644
--- a/src/core/mac/sub_mac.hpp
+++ b/src/core/mac/sub_mac.hpp
@@ -532,6 +532,7 @@ private:
     static void HandleCslTimer(Timer &aTimer);
     void        HandleCslTimer(void);
     void        GetCslWindowEdges(uint32_t &aAhead, uint32_t &aAfter);
+    void        GetNextCslWindowEdges(uint32_t &aAhead, uint32_t &aAfter);
     uint32_t    GetLocalTime(void);
 #if OPENTHREAD_CONFIG_MAC_CSL_DEBUG_ENABLE
     void LogReceived(RxFrame *aFrame);
diff --git a/src/core/mac/sub_mac_csl_receiver.cpp b/src/core/mac/sub_mac_csl_receiver.cpp
index e890687d6..1a6d2e4ee 100644
--- a/src/core/mac/sub_mac_csl_receiver.cpp
+++ b/src/core/mac/sub_mac_csl_receiver.cpp
@@ -178,6 +178,7 @@ void SubMac::HandleCslTimer(void)
      */
     uint32_t periodUs = mCslPeriod * kUsPerTenSymbols;
     uint32_t timeAhead, timeAfter, winStart, winDuration;
+    uint32_t nextTimeAhead, nextTimeAfter;
 
     GetCslWindowEdges(timeAhead, timeAfter);
 
@@ -197,7 +198,8 @@ void SubMac::HandleCslTimer(void)
     {
         if (RadioSupportsReceiveTiming())
         {
-            mCslTimer.FireAt(mCslSampleTime - timeAhead + periodUs);
+            GetNextCslWindowEdges(nextTimeAhead, nextTimeAfter);
+            mCslTimer.FireAt(mCslSampleTime - nextTimeAhead + periodUs);
             timeAhead -= kCslReceiveTimeAhead;
             winStart = mCslSampleTime.GetValue() - timeAhead;
         }
@@ -245,6 +247,24 @@ void SubMac::GetCslWindowEdges(uint32_t &aAhead, uint32_t &aAfter)
     aAfter = Min(semiPeriod, semiWindow + kMinReceiveOnAfter);
 }
 
+void SubMac::GetNextCslWindowEdges(uint32_t &aAhead, uint32_t &aAfter)
+{
+    uint32_t period = mCslPeriod * kUsPerTenSymbols;
+    uint32_t semiPeriod = period / 2;
+    uint32_t curTime, elapsed, semiWindow;
+
+    curTime = GetLocalTime();
+    elapsed = curTime - mCslLastSync.GetValue() + period;
+
+    semiWindow =
+        static_cast<uint32_t>(static_cast<uint64_t>(elapsed) *
+                              (Get<Radio>().GetCslAccuracy() + mCslParentAccuracy.GetClockAccuracy()) / 1000000);
+    semiWindow += mCslParentAccuracy.GetUncertaintyInMicrosec() + Get<Radio>().GetCslUncertainty() * 10;
+
+    aAhead = Min(semiPeriod, semiWindow + kMinReceiveOnAhead + kCslReceiveTimeAhead);
+    aAfter = Min(semiPeriod, semiWindow + kMinReceiveOnAfter);
+}
+
 uint32_t SubMac::GetLocalTime(void)
 {
     uint32_t now;
